### PR TITLE
PHOENIX-5540 Full row index write at the last write phase for immutab…

### PR DESCRIPTION
**Pull request by gokceni**
…le global indexes

### CODE REVIEW

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/index/ImmutableIndexIT.java [hunk 1]
```diff
@@ -417,6 +422,58 @@ public void testGlobalImmutableIndexDelete() throws Exception {
         }
     }
 
+    @Test
+    public void testGlobalImmutableIndexFullWriteOnLast() throws Exception {
```
Comment (priyankporwal) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/index/ImmutableIndexIT.java]: Nit: This test only validates phase1 that it does not write the full row and only unverified. The test name should probably be something like "testGlobalImmutableIndexUnverifiedOnlyInPhase1"

**Comment by BinShi-SecularBird** [comment]
Just curious, the main code change is in MutationState which seems to be generally for both mutable and Immutable global index, so how do we ensure this change is specifically for immutable global index?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/index/ImmutableIndexIT.java (hunk 1 above)
Comment (BinShi-SecularBird) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/index/ImmutableIndexIT.java]: Does the existing test case cover "In the presence of concurrent writes, we skip the last write phase. These writes leave the index writes in unverified status. Similarly, if the first or second phase write fails, we do not proceed with the third phase"?

Comment (gokceni) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/index/ImmutableIndexIT.java]: Other tests cover those cases

File: phoenix-core/src/main/java/org/apache/phoenix/execute/MutationState.java [hunk 2]
```diff
@@ -1239,15 +1239,17 @@ private void filterIndexCheckerMutations(Map<TableInfo, List<Mutation>> mutation
...
+                        // Just send empty with Unverified
+                        Put unverifiedPut = new Put(m.getRow());
+                        unverifiedPut.addColumn(emptyCF, emptyCQ, timestamp, IndexRegionObserver.UNVERIFIED_BYTES);
+                        addToMap(unverifiedIndexMutations, tableInfo, unverifiedPut);
 
                         // Phase 3 mutations are verified
-                        Put verifiedPut = new Put(m.getRow());
-                        verifiedPut.addColumn(emptyCF, emptyCQ, timestamp,
+                        // Send entire mutation with verified
+                        ((Put) m).addColumn(emptyCF, emptyCQ, timestamp,
```
Comment (kadirozde) [inline phoenix-core/src/main/java/org/apache/phoenix/execute/MutationState.java]: The empty column is already part of the mutation at this moment with value "x". Can we remove it first and add the new one as the patch for PHOENIX-5539 does.

File: phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java [hunk 3]
```diff
@@ -506,7 +506,7 @@ private void populatePendingRows(BatchMutateContext context, long now) {
       return mutations;
   }
 
-  public void removeEmptyColumn(Mutation m, byte[] emptyCF, byte[] emptyCQ) {
+  public static void removeEmptyColumn(Mutation m, byte[] emptyCF, byte[] emptyCQ) {
```
Comment (BinShi-SecularBird) [inline phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java]: @gokceni, @kadirozde, IMPO, we should move the following list from IndexRegionObserver to an MutationUtil class in org.apache.phoenix.hbase.index.util or org.apache.phoenix.util:

IndexRegionObserver.UNVERIFIED_BYTES
IndexRegionObserver.VERIFIED_BYTES
IndexRegionObserver.getMaxTimestamp(...)
IndexRegionObserver.removeEmptyColumn(...)

**Comment by gjacoby126** [comment]
JIRA already committed, closing

**Pull request closed without merging**